### PR TITLE
Orion/create release 3

### PR DESCRIPTION
**Pull request by borland**

### CODE REVIEW

File: pkg/question/select.go [hunk 1]
```diff
@@ -28,7 +32,10 @@ func SelectMap[T any](ask Asker, message string, items []T, getKey func(item T)
 	}, &selectedKey); err != nil {
 		return selectedValue, err
 	}
-	selectedValue = optionMap[selectedKey]
+	selectedValue, ok := optionMap[selectedKey]
+	if !ok { // without this explict check SelectMap can return nil, nil which people don't expect
```
Comment (borland) [inline pkg/question/select.go:36]: attn @domenicsim1

File: test/testutil/fakeoctopusserver.go [hunk 2]
```diff
@@ -55,6 +55,8 @@ type MockHttpServer struct {
 	// and go panics and kills the process, so we find out about it, but this is a bit
 	// less confusing to troubleshoot
 	pendingMsgCount int32
+
+	Closed bool
```
Comment (borland) [inline test/testutil/fakeoctopusserver.go:59]: This isn't bulletproof, but adding in the ability to "shut down" the fake HTTP server and fake survey Asker means that in many cases a test failure doesn't deadlock anymore.

Deadlock is not so bad (the test fails, so you go fix it), but the problem is, when go detects a deadlock it hard panics and you don't get to see any of the stdout or stderr so you don't know what happened leading up to the test failure. Usually there's an assert "expected value blah, got blah2" that you'd like to see

File: pkg/cmd/release/create/create.go [hunk 3]
```diff
@@ -171,53 +177,238 @@ func createRun(cmd *cobra.Command, f factory.Factory) error {
...
+		// if there is no git reference specified, ask the server for the list and prompt.
+		// we leave GitCommit alone in interactive mode; we don't prompt, but if it was specified on the
+		// commandline we just pass it through untouched.
+
+		if options.GitReference == "" { // we need a git ref; ask for one
+			gitRef, err := selectGitReference(octopus, asker, spinner, selectedProject)
+			if err != nil {
+				return err
+			}
+			options.GitReference = gitRef.Name // Hold the short name, not the canonical name due to golang url parsing bug replacing %2f with /
```
Comment (borland) [inline pkg/cmd/release/create/create.go:225]: 🤮 but we've got to do it unless we make a breaking change to the api client to get rid of sling

File: pkg/cmd/release/create/create.go [hunk 4]
```diff
@@ -171,53 +177,238 @@ func createRun(cmd *cobra.Command, f factory.Factory) error {
...
 	}
 	options.ChannelName = selectedChannel.Name
-
 	if err != nil {
 		return err
 	}
 
+	// TODO package version prompting goes here BEFORE specification of the release version
+
 	if options.Version == "" {
```
Comment (borland) [inline pkg/cmd/release/create/create.go:275]: guessing the next release version is really hard!

File: pkg/cmd/release/create/create.go [hunk 5]
```diff
@@ -171,53 +177,238 @@ func createRun(cmd *cobra.Command, f factory.Factory) error {
...
+			if err == nil && (pkgVersions.Items != nil && len(pkgVersions.Items) == 0) {
+				err = errors.New("could not find any versions for package")
+			}
+			if err != nil {
+				spinner.Stop()
+				return err
+			}
+
+			tagFilters := make([]*regexp.Regexp, 0, len(selectedChannel.Rules))
+			// TODO version range filters; we need to parse either Nuget or Maven syntax, which is super hard. Is this the right way to go about it?
```
Comment (borland) [inline pkg/cmd/release/create/create.go:354]: I have concerns here. Would be good to discuss

File: pkg/cmd/release/create/create.go [hunk 6]
```diff
@@ -409,20 +598,38 @@ func selectPackageOverrides(octopus *octopusApiClient.Client, ask question.Asker
...
-			}
-		}
+		return p.Name
+	})
+}
 
-		return ""
+func selectGitReference(octopus *octopusApiClient.Client, ask question.Asker, spinner factory.Spinner, project *projects.Project) (*projects.GitReference, error) {
+	spinner.Start()
+	branches, err := octopus.Projects.GetGitBranches(project)
```
Comment (borland) [inline pkg/cmd/release/create/create.go:616]: This is the dangerous part here; there could potentially be hundreds or even thousands of branches and tags in a large repo. This just dumbly loads them all

File: pkg/cmd/release/create/create_test.go [hunk 7]
```diff
@@ -120,65 +123,242 @@ func TestReleaseCreate_AskQuestions(t *testing.T) {
...
 
 		qa.ExpectQuestion(t, &survey.Select{
 			Message: "Select the project in which the release will be created",
-			Options: []string{"Fire Project"},
-		}).AnswerWith("Fire Project")
+			Options: []string{project.Name},
+		}).AnswerWith(project.Name)
+
+		// CLI will load all the branches and tags
+		api.ExpectRequest(t, "GET", "/api/Spaces-1/projects/"+projectID+"/git/branches").RespondWith(resources.Resources[*projects.GitReference]{
```
Comment (borland) [inline pkg/cmd/release/create/create_test.go:167]: for discussion: FYI @slewis74 - our unit tests are currently coupled to the internal workings of the API client, which is not great

**Review by domenicsim1 — APPROVED** [review]

**Pull request merged**